qsimcirq/qsim_simulator.py

```diff
@@ -435,12 +435,67 @@ def compute_amplitudes_sweep_iter(
             options["s"] = self.get_seed()
             yield simulator_fn(options)
 
+    def simulate(
+        self,
+        program: cirq.AbstractCircuit,
+        param_resolver: cirq.ParamResolverOrSimilarType = None,
+        qubit_order: cirq.QubitOrderOrList = cirq.ops.QubitOrder.DEFAULT,
+        initial_state: Any = None,
+        as_1d_state_vector: bool = False,
+    ) -> cirq.sim.simulator.TSimulationTrialResult:
+        """Simulates the supplied Circuit.
+
+        This method returns a result which allows access to the entire
+        simulator's final state.
+
+        Args:
+            program: The circuit to simulate.
+            param_resolver: Parameters to run with the program.
+            qubit_order: Determines the canonical ordering of the qubits. This
+                is often used in specifying the initial state, i.e. the
+                ordering of the computational basis states.
+            initial_state: The initial state for the simulation. The form of
+                this state depends on the simulation implementation. See
+                documentation of the implementing class for details.
+            as_1d_state_vector: Whether the returned state vector is 1D or
+                has number of dimensions equal number of qubits.
+                Operations on 1D representation are significantly slower than
+                the other representation and might not even work.
+                The 1D representation should only be used when the number of qubits
+                is larger than the dimension limit on numpy arrays (numpy/numpy#5744).
+
+        Returns:
+            SimulationTrialResults for the simulation. Includes the final state.
+        """
+        return self.simulate_sweep(
+            program,
+            cirq.study.ParamResolver(param_resolver),
+            qubit_order,
+            initial_state,
+            as_1d_state_vector,
+        )[0]
+
+    def simulate_sweep(
+        self,
+        program: cirq.AbstractCircuit,
+        params: cirq.Sweepable,
+        qubit_order: cirq.QubitOrderOrList = cirq.ops.QubitOrder.DEFAULT,
+        initial_state: Any = None,
+        as_1d_state_vector: bool = False,
+    ) -> List[cirq.sim.simulator.TSimulationTrialResult]:
+        return list(
+            self.simulate_sweep_iter(
+                program, params, qubit_order, initial_state, as_1d_state_vector
+            )
+        )
+
     def simulate_sweep_iter(
         self,
         program: cirq.Circuit,
         params: cirq.Sweepable,
         qubit_order: cirq.QubitOrderOrList = cirq.QubitOrder.DEFAULT,
         initial_state: Optional[Union[int, np.ndarray]] = None,
+        as_1d_state_vector: bool = False,
```

Comment:
This violates the API defined by [`cirq.SimulatesFinalState.simulate_sweep_iter`](https://github.com/quantumlib/Cirq/blob/8e4e7d147d6d4ea5dd8111ba21a715aa3acd955c/cirq-core/cirq/sim/simulator.py#L525). If there are plans to modify that function as well, please link the relevant Cirq PR. (The required `cirq` version for qsim will also need to be updated if this is the case.)

Reply:
it doesn't violate the API. but it changes the internal data representation just for this  if and only if `as_1d_state_vector = True` which by default is `False` so nothing changes unless the caller explicitly wants the 1D representation. This is done only for the qsim simulator.

The 1D representation is used for only one reason and that is to report the result, because the normal representation is a tensor that has number of dimensions equal to the `num_qubits` which breaks when the number of qubits is greater than the limit on numpy array dimensions (see issue in docstring & PR description)

Reply:
`QSimSimulator` inherits from `cirq.SimulatesFinalState`, whose `simulate_sweep_iter` method does not have this new argument. Even though the implementation here can accept any valid input to the function of the parent class, it's still a violation of the API.

Reply:
I changed the approach to adding a new method rather than extending the existing API.